### PR TITLE
Add filter() fn, update pagination() fn

### PR DESCRIPTION
**Pull request by Wulf**

### CODE REVIEW

File: src/bin/main.rs
Comment (hasezoey) [inline src/bin/main.rs:107]:
```suggestion
    /// Diesel provides the following backends:
```

File: src/bin/main.rs
Comment (hasezoey) [inline src/bin/main.rs:115]: i think for this we should not assume any default backend

File: src/code.rs
Comment (hasezoey) [inline src/code.rs:426]: why update the parameter names with `param_`?

File: src/code.rs
Comment (hasezoey) [inline src/code.rs:429]:
```suggestion
        let param_page = param_page_starting_with_1.max(1);
```

guessing by the previous implementation and the updated doc-comment, shouldnt the lowest be 1 instead of 0?

File: src/code.rs
Comment (hasezoey) [inline src/code.rs:459]: this looks quite redundant, is there not better option?

File: src/code.rs
Comment (hasezoey) [inline src/code.rs:447]:
```suggestion
        let diesel_backend = &config.diesel_backend;
```

i am pretty sure this does not need to be cloned here

File: src/code.rs
Comment (hasezoey) [inline src/code.rs:466]: this should probably be removed if there is no example (and be put up as a issue)

File: src/code.rs
Comment (hasezoey) [inline src/code.rs:528]: i am pretty sure this will lead to errors (because int conversion is forgotten), instead use the helper type by first converting the column to a `StructField` like `StructField::from(column)`

File: src/code.rs
Comment (hasezoey) [inline src/code.rs:544]:
```suggestion
#[derive(Debug, Clone)]
```
the debug derive should always be there

File: src/code.rs
Comment (hasezoey) [inline src/code.rs:545]: this should probably be a new structtype

File: src/code.rs
Comment (hasezoey) [inline src/code.rs:473]:
```suggestion
        {filters}
```

this otherwise causes double-newlines in output.
alternative is to remove the newline in the map-format (but would have to adjust for each loop)

**Review by hasezoey — CHANGES_REQUESTED** [review]
There are some changes that should likely be done.

i personally dont really like having to add such verbose "query = query" things, but there is no helper trait for that in diesel, right?

File: src/bin/main.rs
Comment (Wulf) [reply src/bin/main.rs:115]: done

File: src/code.rs
Comment (Wulf) [reply src/code.rs:426]: it's in-line with the rest of the generated code, but I'll remove this :)

File: src/code.rs
Comment (Wulf) [reply src/code.rs:429]: ah, this is actually a bug, it should be:
`let param_page = (page_starting_with_1 - 1).max(0);`

Comment (Wulf) [reply src/code.rs:429]: ok, changed this: I reverted to a 0-based page API

File: src/code.rs
Comment (Wulf) [reply src/code.rs:447]: Done!

File: src/code.rs
Comment (Wulf) [reply src/code.rs:466]: added an example!

File: src/code.rs
Comment (Wulf) [reply src/code.rs:459]: feel free to suggest changes. This new filter() API is an experiment I want to run and play with. We might want to remove it in the future as intellisense gets better and better around macros.

File: src/code.rs
Comment (Wulf) [reply src/code.rs:528]: I thought about leveraging StructField, but I remember there were some things I bumped into that made me change my mind and took me towards this path. Let's merge this in and discuss any errors as they come up.

Things that StructField deal with (like "str/String/Cow", etc.) don't apply to this filter API

Comment (Wulf) [reply src/code.rs:528]: let's merge this in for now and deal with any issues that come up

File: src/code.rs
Comment (Wulf) [reply src/code.rs:545]: let's leave this for now as it's an experimental API that needs more iterations before we solidify it as a core struct.

File: src/code.rs
Comment (hasezoey) [reply src/code.rs]: to my understanding the style (implicitly) was to add the `param_` prefix to struct values as parameters, not separate options (like page size)

File: src/code.rs
Comment (hasezoey) [reply src/code.rs:537]: see #65, #67 which i cleaned-up with #85

this would be required because `column` in this case is a `ParsedColumnMacro`, which still means it is "raw", which would look like this:
```txt
ParsedColumnMacro {
    ty: "i32",
    name: Ident(
        unsigned_nullable,
    ),
    is_nullable: true,
    is_unsigned: true,
}
```
so you would need to convert the `ty` `i32` to `u32` and also make it wrapped in `Options`, and `StructField` was added in #85 exactly to deduplicate this code to not have issues like this anymore

File: src/lib.rs
Comment (hasezoey) [inline src/lib.rs:343]: why have backend validation against a specific list now when

> `- or, your custom diesel backend type (struct which implements `diesel::backend::Backend`)`

is still said and possible?

File: src/code.rs
Comment (hasezoey) [reply src/code.rs:459]: i had already looked, but there does not seem to be a diesel helper trait which would have a Update-like struct be the filter, something as simple as `.filter(FilterStruct { someprop: Some(SomeValue), ..Default::default() })` similar to what `.select(Struct::as_select())` would look like

File: src/code.rs
Comment (hasezoey) [inline src/code.rs:473]:
```suggestion
    ///     let query = Todo::filter(TodoFilter {
    ///       completed: Some(true),
    ///       ..Default::default()
    ///     });
```

did i miss something or why are there `{{` and `}}`?

File: src/code.rs
Comment (hasezoey) [inline src/code.rs:558]: if all fields are a option anyway, why not just derive Default? (because option defaults to `None`)

**Review by hasezoey — COMMENTED** [review]
Some more comments.

also noticed that the tests have not been re-run since 7518fa18911cd4b836fc8888c6a177b967bef21b

File: src/lib.rs
Comment (Wulf) [reply src/lib.rs:343]: thank you, good catch!

File: src/code.rs
Comment (Wulf) [reply src/code.rs:473]: I think it's because my editor was being weird about it haha
image

Comment (Wulf) [reply src/code.rs:473]: ah nevermind, it's because it was in a `format!()`. Here I was thinking why this simple doc comment wasn't working!

In format!(), you need to escape '{' usage by using two of them. Hence the {{ }}. We do this a lot in dsync!

**Comment by Wulf** [comment]
Thanks for your review @hasezoey, I've taken your feedback and applied it! 🙌

File: src/code.rs
Comment (hasezoey) [reply src/code.rs]: right, also forgot it was inside a `format!()`, i was just skimming through with the github editor (no in-detail highlighting or playing around with some changes)

File: src/lib.rs
Comment (hasezoey) [inline src/lib.rs:337]: does that really need `.clone`?

Comment (Wulf) [reply src/lib.rs]: removed :)

File: src/code.rs
Comment (hasezoey) [inline src/code.rs:446]: this block should likely be its own function to keep the functions smaller, but this can be delayed until we refactor this (see #105)

**Review by hasezoey — CHANGES_REQUESTED** [review]
I dont really like how the current filter implementation works (and i dont know a improvement), but the code otherwise looks fine, some points that could probably be changed though:
- the validate config message will look somewhat weird, because the only case it currently will be triggered is by the input being empty
- the example for the `filter` method could make use of the table names instead of the generic `Todo`

it would be nice if we could somehow consistently figure out what backend is being used from the connection string, and only require it if we cant figure it out (this should be possible for sqlite, mysql and postgres, but i dont know about clap liking it)

Marking this still as "Request changes" because this PR will need to be rebased onto latest master

File: src/code.rs
Comment (Wulf) [reply src/code.rs:446]: Awesome, I’ll take a look at this later and try to rebase 🙌

Comment (hasezoey) [reply src/code.rs:446]: unless you want to delay this PR, i will hold off on merging other stuff (even if approved) so you can get this (and possibly your other PR's) merged without having to rebase it often

Comment (Wulf) [reply src/code.rs:446]: Thanks @hasezoey, I really appreciate your consideration 🤗 

Please do approve this PR if you think it's generally acceptable. That way, I can rebase and merge later without having to stop any others.

Comment (hasezoey) [reply src/code.rs:446]: > Please do approve this PR if you think it's generally acceptable. That way, I can rebase and merge later without having to stop any others.

i take this as meaning you have not rebased yet, so i will merge the other standing (approved) PRs to not have to wait, after that i will approve

**Review by hasezoey — APPROVED** [review]
Approving as per https://github.com/Wulf/dsync/pull/103#discussion_r1388743776 to be able to rebase and merge it directly.

unless this PR will get closed because of #120

**Comment by Wulf** [comment]
Also going to close this -- we can reference it later :-)

**Pull request closed without merging**
